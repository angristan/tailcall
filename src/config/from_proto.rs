#![allow(dead_code)] // TODO check what to do..

use std::collections::BTreeMap;

use convert_case::{Case, Casing};
use prost_reflect::prost_types::{
    DescriptorProto, EnumDescriptorProto, FileDescriptorSet, ServiceDescriptorProto,
};

use crate::config::{Arg, Config, Field, Type};

fn convert_ty(proto_ty: &str) -> String {
    let binding = proto_ty.to_lowercase();
    let proto_ty = binding.strip_prefix("type_").unwrap_or(proto_ty);
    match proto_ty {
        "double" | "float" => "Float",
        "int32" | "int64" | "fixed32" | "fixed64" | "uint32" | "uint64" => "Int",
        "bool" => "Boolean",
        "string" | "bytes" => "String",
        x => x,
    }
    .to_string()
}

fn get_output_ty(output_ty: &str) -> (String, bool) {
    // type, required
    match output_ty {
        "google.protobuf.Empty" => {
            ("String".to_string(), false) // If it's no response is expected, we
                                          // return a nullable string type
        }
        any => (any.to_string(), true), /* Setting it not null by default. There's no way to
                                         * infer this from proto file */
    }
}

fn get_arg(input_ty: &str) -> Option<(String, Arg)> {
    match input_ty {
        "google.protobuf.Empty" | "" => None,
        any => {
            let key = convert_ty(any).to_case(Case::Camel);
            let val = Arg {
                type_of: any.to_string(),
                list: false,
                required: true,
                /* Setting it not null by default. There's no way to infer this
                 * from proto file */
                doc: None,
                modify: None,
                default_value: None,
            };

            Some((key, val))
        }
    }
}

fn append_enums(map: &mut BTreeMap<String, Type>, enums: Vec<EnumDescriptorProto>) {
    for enum_ in enums {
        let ty = Type {
            variants: Some(enum_.value.iter().map(|v| v.name().to_string()).collect()),
            ..Default::default()
        };
        map.insert(enum_.name().to_string(), ty);
    }
}

fn append_msg_type(map: &mut BTreeMap<String, Type>, messages: Vec<DescriptorProto>) {
    if messages.is_empty() {
        return;
    }
    for message in messages {
        let msg_name = message.name().to_string();
        append_enums(map, message.enum_type);
        append_msg_type(map, message.nested_type);

        let mut ty = Type::default();

        for field in message.field {
            let field_name = field.name().to_string();
            let mut cfg_field = Field::default();

            let label = field.label().as_str_name().to_lowercase();
            cfg_field.list = label.contains("repeated");

            if field.r#type.is_some() {
                // for non-primitive types
                let type_of = convert_ty(field.r#type().as_str_name());
                cfg_field.type_of = type_of.to_string();
            } else {
                cfg_field.type_of = convert_ty(field.type_name());
            }

            ty.fields.insert(field_name, cfg_field);
        }
        map.insert(msg_name, ty);
    }
}

fn append_service(
    map: &mut BTreeMap<String, Type>,
    services: Vec<ServiceDescriptorProto>,
    query: String,
) {
    if services.is_empty() {
        return;
    }
    let mut ty = Type::default();

    for service in services {
        for method in service.method {
            let mut cfg_field = Field::default();
            if let Some((k, v)) = get_arg(method.input_type()) {
                cfg_field.args.insert(k, v);
            }

            let (output_ty, required) = get_output_ty(method.output_type());
            cfg_field.type_of = output_ty;
            cfg_field.required = required;
            let name = method.name().to_case(Case::Camel);
            ty.fields.insert(name, cfg_field);
        }
    }
    map.insert(query, ty);
}

pub fn from_proto(
    descriptor_set: FileDescriptorSet,
    query: Option<String>,
) -> anyhow::Result<Config> {
    let mut config = Config::default();
    let mut types = BTreeMap::new();
    let query = query.unwrap_or("Query".to_string());

    for file_descriptor in descriptor_set.file {
        append_enums(&mut types, file_descriptor.enum_type);
        append_msg_type(&mut types, file_descriptor.message_type);
        append_service(&mut types, file_descriptor.service, query.clone());
    }

    config.types = types;

    Ok(config)
}

#[cfg(test)]
mod test { // TODO add proper tests
    use prost_reflect::prost_types::FileDescriptorSet;

    use crate::config::from_proto::from_proto;

    static FOO: &str = r#"
    syntax = "proto3";

        import "google/protobuf/empty.proto";

        package news;

        enum Status {
          PUBLISHED = 0;
          DRAFT = 1;
        }


        message News {
          int32 id = 1;
          string title = 2;
          string body = 3;
          string postImage = 4;
          Status foo = 5;
        }

        service NewsService {
          rpc GetAllNews (google.protobuf.Empty) returns (NewsList) {}
          rpc GetNews (NewsId) returns (News) {}
          rpc GetMultipleNews (MultipleNewsId) returns (NewsList) {}
          rpc DeleteNews (NewsId) returns (google.protobuf.Empty) {}
          rpc EditNews (News) returns (News) {}
          rpc AddNews (News) returns (News) {}
        }

        message NewsId {
          int32 id = 1;
        }

        message MultipleNewsId {
          repeated NewsId ids = 1;
        }

        message NewsList {
          repeated News news = 1;
        }
    "#;

    #[test]
    fn test_from_proto() -> anyhow::Result<()> {
        let mut set = FileDescriptorSet::default();
        let file_desc = protox_parse::parse("news.proto", FOO)?;
        set.file.push(file_desc);

        let result = from_proto(set, None)?;
        println!("{}", result.to_sdl());
        Ok(())
    }
}
