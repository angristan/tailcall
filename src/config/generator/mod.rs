pub mod from_proto;
mod proto_generator;
