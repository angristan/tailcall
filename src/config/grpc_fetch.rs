use anyhow::{Context, Result};
use base64::prelude::BASE64_STANDARD;
use base64::Engine;
use hyper::header::HeaderName;
use nom::AsBytes;
use prost::Message;
use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::grpc::protobuf::{ProtobufOperation, ProtobufSet};
use crate::grpc::RequestTemplate;
use crate::mustache::Mustache;
use crate::runtime::{TargetRuntime, TargetRuntimeContext};

const REFLECTION_PROTO: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/proto/reflection.proto"
));

/// This function is just used for better exception handling
fn get_protobuf_set() -> Result<ProtobufSet> {
    let descriptor = protox_parse::parse("reflection", REFLECTION_PROTO)?;
    let mut descriptor_set = FileDescriptorSet::default();
    descriptor_set.file.push(descriptor);
    ProtobufSet::from_proto_file(&descriptor_set)
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
struct OriginalRequest {
    list_services: Option<String>,
    file_containing_symbol: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct Service {
    name: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct ListServicesResponse {
    service: Vec<Service>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
struct FileDescriptorProtoResponse {
    file_descriptor_proto: Vec<String>,
}

impl FileDescriptorProtoResponse {
    fn get(self) -> Result<Vec<u8>> {
        let file_descriptor_proto = self
            .file_descriptor_proto
            .first()
            .context("Received empty fileDescriptorProto")?;

        Ok(BASE64_STANDARD.decode(file_descriptor_proto)?)
    }
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
struct CustomResponse {
    original_request: OriginalRequest,
    list_services_response: Option<ListServicesResponse>,
    file_descriptor_response: Option<FileDescriptorProtoResponse>,
}

/// Makes `ListService` request to the grpc reflection server
pub async fn list_all_files(url: &str, target_runtime: &TargetRuntime) -> Result<Vec<String>> {
    let protobuf_set = get_protobuf_set()?;

    let reflection_service =
        protobuf_set.find_service("grpc.reflection.v1alpha.ServerReflection")?;
    let operation = reflection_service.find_operation("ServerReflectionInfo")?;

    // let mut methods = vec![];
    let mut url: url::Url = url.parse()?;
    url.set_path("grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo");

    let req_template = RequestTemplate {
        url: Mustache::parse(url.as_str())?,
        headers: vec![(
            HeaderName::from_static("content-type"),
            Mustache::parse("application/grpc+proto")?,
        )],
        body: Mustache::parse(json!({"list_services": ""}).to_string().as_str()).ok(),
        operation: operation.clone(),
        operation_type: Default::default(),
    };

    let ctx = TargetRuntimeContext {
        runtime: target_runtime,
        vars: &Default::default(),
        headers: Default::default(),
    };

    let req = req_template.render(&ctx)?.to_request()?;

    let resp = target_runtime.http.execute(req).await?;
    let body = resp.body.as_bytes();

    let response: Value = serde_json::from_value(operation.convert_output(body)?.into_json()?)?;
    let response: CustomResponse = serde_json::from_value(response)?;

    // Extracting names from services
    let methods: Vec<String> = response
        .list_services_response
        .context("Expected listServicesResponse but found none")?
        .service
        .iter()
        .map(|s| s.name.clone())
        .collect();

    Ok(methods)
}

/// Makes `Get Service` request to the grpc reflection server
pub async fn get_by_service(
    url: &str,
    target_runtime: &TargetRuntime,
    service: &str,
) -> Result<FileDescriptorProto> {
    let protobuf_set = get_protobuf_set()?;

    let reflection_service =
        protobuf_set.find_service("grpc.reflection.v1alpha.ServerReflection")?;
    let operation = reflection_service.find_operation("ServerReflectionInfo")?;

    let mut url: url::Url = url.parse()?;
    url.set_path("grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo");

    let req_template = RequestTemplate {
        url: Mustache::parse(url.as_str())?,
        headers: vec![(
            HeaderName::from_static("content-type"),
            Mustache::parse("application/grpc+proto")?,
        )],
        body: Mustache::parse(
            json!({"file_containing_symbol": service})
                .to_string()
                .as_str(),
        )
        .ok(),
        operation: operation.clone(),
        operation_type: Default::default(),
    };

    let ctx = TargetRuntimeContext {
        runtime: target_runtime,
        vars: &Default::default(),
        headers: Default::default(),
    };

    let req = req_template.render(&ctx)?.to_request()?;

    request_proto(req, target_runtime, operation).await
}

/// Makes `Get Proto/Symbol Name` request to the grpc reflection server
pub async fn get_by_proto_name(
    url: &str,
    target_runtime: &TargetRuntime,
    proto_name: &str,
) -> Result<FileDescriptorProto> {
    let protobuf_set = get_protobuf_set()?;

    let reflection_service =
        protobuf_set.find_service("grpc.reflection.v1alpha.ServerReflection")?;
    let operation = reflection_service.find_operation("ServerReflectionInfo")?;

    let mut url: url::Url = url.parse()?;
    url.set_path("grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo");

    let req_template = RequestTemplate {
        url: Mustache::parse(url.as_str())?,
        headers: vec![(
            HeaderName::from_static("content-type"),
            Mustache::parse("application/grpc+proto")?,
        )],
        body: Mustache::parse(json!({"file_by_filename": proto_name}).to_string().as_str()).ok(),
        operation: operation.clone(),
        operation_type: Default::default(),
    };

    let ctx = TargetRuntimeContext {
        runtime: target_runtime,
        vars: &Default::default(),
        headers: Default::default(),
    };

    let req = req_template.render(&ctx)?.to_request()?;

    request_proto(req, target_runtime, operation).await
}

async fn request_proto(
    req: reqwest::Request,
    target_runtime: &TargetRuntime,
    operation: ProtobufOperation,
) -> Result<FileDescriptorProto> {
    let resp = target_runtime.http.execute(req).await?;
    let body = resp.body.as_bytes();

    let response: CustomResponse =
        serde_json::from_value(operation.convert_output(body)?.into_json()?)?;
    let file_descriptor_resp = response
        .file_descriptor_response
        .context("Expected fileDescriptorResponse but found none")?;
    let file_descriptor_proto =
        FileDescriptorProto::decode(file_descriptor_resp.get()?.as_bytes())?;

    Ok(file_descriptor_proto)
}

#[cfg(test)]
mod grpc_fetch {
    use std::path::PathBuf;

    use anyhow::Result;

    use crate::config::grpc_fetch::{get_by_proto_name, get_by_service, list_all_files};

    const NEWS_PROTO: &[u8] = include_bytes!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/tests/fake_descriptor.bin"
    ));

    const REFLECTION_LIST_ALL: &[u8] = &[
        0, 0, 0, 0, 70, 18, 2, 58, 0, 50, 64, 10, 18, 10, 16, 110, 101, 119, 115, 46, 78, 101, 119,
        115, 83, 101, 114, 118, 105, 99, 101, 10, 42, 10, 40, 103, 114, 112, 99, 46, 114, 101, 102,
        108, 101, 99, 116, 105, 111, 110, 46, 118, 49, 97, 108, 112, 104, 97, 46, 83, 101, 114,
        118, 101, 114, 82, 101, 102, 108, 101, 99, 116, 105, 111, 110,
    ];

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }
    #[tokio::test]
    async fn test_resp_file_name() -> Result<()> {
        let server = start_mock_server();

        let http_reflection_file_mock = server.mock(|when, then| {
            when.method(httpmock::Method::POST)
                .path("/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo")
                .body("\0\0\0\0\x0c\x1a\nnews.proto");
            then.status(200).body(NEWS_PROTO);
        });

        let runtime = crate::runtime::test::init(None);
        let resp = get_by_proto_name(
            &format!("http://localhost:{}", server.port()),
            &runtime,
            "news.proto",
        )
        .await?;
        let mut news_proto = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        news_proto.push("src/grpc/tests/news.proto");
        let content = runtime.file.read(news_proto.to_str().unwrap()).await?;
        let expected = protox_parse::parse("news.proto", &content)?;

        assert_eq!(expected.name(), resp.name());

        http_reflection_file_mock.assert();
        Ok(())
    }

    #[tokio::test]
    async fn test_resp_service() -> Result<()> {
        let server = start_mock_server();

        let http_reflection_file_mock = server.mock(|when, then| {
            when.method(httpmock::Method::POST)
                .path("/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo")
                .body("\0\0\0\0\x12\"\x10news.NewsService");
            then.status(200).body(NEWS_PROTO);
        });

        let runtime = crate::runtime::test::init(None);
        let resp = get_by_service(
            &format!("http://localhost:{}", server.port()),
            &runtime,
            "news.NewsService",
        )
        .await?;
        let mut news_proto = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        news_proto.push("src/grpc/tests/news.proto");
        let content = runtime.file.read(news_proto.to_str().unwrap()).await?;
        let expected = protox_parse::parse("news.proto", &content)?;

        assert_eq!(expected.name(), resp.name());

        http_reflection_file_mock.assert();
        Ok(())
    }

    #[tokio::test]
    async fn test_resp_list_all() -> Result<()> {
        let server = start_mock_server();

        let http_reflection_list_all = server.mock(|when, then| {
            when.method(httpmock::Method::POST)
                .path("/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo")
                .body("\0\0\0\0\x02:\0");
            then.status(200).body(REFLECTION_LIST_ALL);
        });

        let runtime = crate::runtime::test::init(None);
        let resp = list_all_files(&format!("http://localhost:{}", server.port()), &runtime).await?;

        assert_eq!(
            [
                "news.NewsService".to_string(),
                "grpc.reflection.v1alpha.ServerReflection".to_string()
            ]
            .to_vec(),
            resp
        );

        http_reflection_list_all.assert();

        Ok(())
    }
}
