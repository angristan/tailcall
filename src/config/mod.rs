pub use config::*;
pub use config_module::*;
pub(crate) use expr::*;
pub(crate) use group_by::*;
pub(crate) use key_values::*;
pub(crate) use link::*;
pub(crate) use reader::*;
pub(crate) use server::*;
pub use source::*;
pub use upstream::*;
mod config;
mod config_module;
mod expr;
mod from_document;
mod group_by;
mod into_document;
mod key_values;
mod link;
mod n_plus_one;
mod reader;
mod server;
mod source;
mod upstream;
