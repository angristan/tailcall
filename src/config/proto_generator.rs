use derive_setters::Setters;

#[derive(Setters)]
pub struct ProtoGeneratorConfig {
    query: String,
    mutation: String,
    is_mutation_fxn: Box<dyn Fn(&str) -> bool>,
}

impl ProtoGeneratorConfig {
    pub fn new(
        query: Option<String>,
        mutation: Option<String>,
        is_mutation_fxn: Box<dyn Fn(&str) -> bool>,
    ) -> Self {
        Self {
            query: query.unwrap_or("Query".to_string()),
            mutation: mutation.unwrap_or("Mutation".to_string()),
            is_mutation_fxn,
        }
    }

    pub fn is_mutation(&self, name: &str) -> bool {
        (self.is_mutation_fxn)(name)
    }
    pub fn get_query(&self) -> &str {
        self.query.as_str()
    }

    pub fn get_mutation(&self) -> &str {
        self.mutation.as_str()
    }
}

impl Default for ProtoGeneratorConfig {
    fn default() -> Self {
        Self {
            query: "Query".to_string(),
            mutation: "Mutation".to_string(),
            is_mutation_fxn: Box::new(|_| false),
        }
    }
}
