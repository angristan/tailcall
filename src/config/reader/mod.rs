pub use grpc_fetch::*;
pub use reader::*;
mod grpc_fetch;
mod reader;
