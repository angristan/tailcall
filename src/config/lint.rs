use std::collections::{BTreeMap, BTreeSet};

use convert_case::Casing;
use serde::{Deserialize, Serialize};

use crate::config::{AddField, Config, Type};

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema)]
pub struct Lint {
    #[serde(rename = "field")]
    pub field_lint: bool,
    #[serde(rename = "type")]
    pub type_lint: bool,
    #[serde(rename = "enum")]
    pub enum_lint: bool,
    #[serde(rename = "enumValue")]
    pub enum_value_lint: bool,
    #[serde(rename = "autoFix")]
    pub autofix: bool,
}

// Type -> PascalCase
// Field -> camelCase
// Enum -> PascalCase
// Enum Values -> ALL_CAPS

pub fn lint_fix(mut config: Config) -> Config {
    if let Some(lint) = &config.server.lint {
        let mut fixed_types = BTreeMap::new();
        for (k, ty) in config.types.iter() {
            fixed_types.insert(
                match lint.type_lint {
                    true => k.to_case(convert_case::Case::Pascal),
                    false => k.clone(),
                },
                fix_type(lint, ty),
            );
        }
        config.types = fixed_types;
    }
    config
}

// TODO: check variants, fields and added_fields
fn fix_type(lint: &Lint, ty: &Type) -> Type {
    let mut new_ty = Type::default();

    // for enum
    if lint.enum_lint {
        if let Some(variants) = &ty.variants {
            let mut set = BTreeSet::new();
            for variant in variants {
                set.insert(variant.to_case(convert_case::Case::Upper));
            }
            new_ty.variants = Some(set);
        }
    }

    if lint.field_lint {
        for (k, field) in &ty.fields {
            new_ty
                .fields
                .insert(k.to_case(convert_case::Case::Camel), field.clone());
        }
        for field in &ty.added_fields {
            new_ty.added_fields.push(AddField {
                name: field.name.to_case(convert_case::Case::Camel),
                path: field.path.clone(),
            })
        }
    }

    new_ty
}
