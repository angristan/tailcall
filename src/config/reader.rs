use std::collections::{HashMap, VecDeque};

use anyhow::Context;
use futures_util::future::join_all;
use futures_util::TryFutureExt;
use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};
use protox::file::{FileResolver, GoogleFileResolver};
use url::Url;

use super::{ConfigSet, ExprBody, Extensions, Script, ScriptOptions};
use crate::config::{Config, Source};
use crate::target_runtime::TargetRuntime;

const NULL_STR: &str = "\0\0\0\0\0\0\0";

/// Reads the configuration from a file or from an HTTP URL and resolves all linked extensions to create a ConfigSet.
pub struct ConfigReader {
    runtime: TargetRuntime,
}

/// Response of a file read operation
struct FileRead {
    content: String,
    path: String,
}

impl ConfigReader {
    pub fn init(runtime: TargetRuntime) -> Self {
        Self { runtime }
    }

    /// Reads a file from the filesystem or from an HTTP URL
    async fn read_file<T: ToString>(&self, file: T) -> anyhow::Result<FileRead> {
        // Is an HTTP URL
        let content = if let Ok(url) = Url::parse(&file.to_string()) {
            let response = self
                .runtime
                .http
                .execute(reqwest::Request::new(reqwest::Method::GET, url))
                .await?;

            String::from_utf8(response.body.to_vec())?
        } else {
            // Is a file path

            self.runtime.file.read(&file.to_string()).await?
        };

        Ok(FileRead { content, path: file.to_string() })
    }

    /// Reads all the files in parallel
    async fn read_files<T: ToString>(&self, files: &[T]) -> anyhow::Result<Vec<FileRead>> {
        let files = files.iter().map(|x| {
            self.read_file(x.to_string())
                .map_err(|e| e.context(x.to_string()))
        });
        let content = join_all(files)
            .await
            .into_iter()
            .collect::<anyhow::Result<Vec<_>>>()?;
        Ok(content)
    }

    /// Reads the script file and replaces the path with the content
    async fn ext_script(&self, mut config_set: ConfigSet) -> anyhow::Result<ConfigSet> {
        let config = &mut config_set.config;
        if let Some(Script::Path(ref options)) = &config.server.script {
            let timeout = options.timeout;
            let script = self.read_file(options.src.clone()).await?.content;
            config.server.script = Some(Script::File(ScriptOptions { src: script, timeout }));
        }
        Ok(config_set)
    }

    /// Reads a single file and returns the config
    pub async fn read<T: ToString>(&self, file: T) -> anyhow::Result<ConfigSet> {
        self.read_all(&[file]).await
    }

    /// Reads all the files and returns a merged config
    pub async fn read_all<T: ToString>(&self, files: &[T]) -> anyhow::Result<ConfigSet> {
        let files = self.read_files(files).await?;
        let mut config_set = ConfigSet::default();

        for file in files.iter() {
            let source = Source::detect(&file.path)?;
            let schema = &file.content;

            // Create initial config set
            let new_config_set = self.resolve(Config::from_source(source, schema)?).await?;

            // Merge it with the original config set
            config_set = config_set.merge_right(&new_config_set);
        }
        Ok(config_set)
    }

    /// Resolves all the links in a Config to create a ConfigSet
    pub async fn resolve(&self, config: Config) -> anyhow::Result<ConfigSet> {
        // Create initial config set
        let config_set = ConfigSet::from(config);

        // Extend it with the worker script
        let config_set = self.ext_script(config_set).await?;

        // Extend it with protobuf definitions for GRPC
        let config_set = self.ext_grpc(config_set).await?;

        Ok(config_set)
    }

    /// Returns final ConfigSet from Config
    pub async fn ext_grpc(&self, mut config_set: ConfigSet) -> anyhow::Result<ConfigSet> {
        let config = &config_set.config;
        let mut descriptors: HashMap<String, FileDescriptorProto> = HashMap::new();
        let mut grpc_file_descriptor = FileDescriptorSet::default();
        for (_, typ) in config.types.iter() {
            for (_, fld) in typ.fields.iter() {
                let proto_path = if let Some(grpc) = &fld.grpc {
                    &grpc.proto_path
                } else if let Some(ExprBody::Grpc(grpc)) = fld.expr.as_ref().map(|e| &e.body) {
                    &grpc.proto_path
                } else {
                    NULL_STR
                };

                if proto_path != NULL_STR {
                    descriptors = self
                        .resolve_descriptors(descriptors, proto_path.to_string())
                        .await?;
                }
            }
        }
        for (_, v) in descriptors {
            grpc_file_descriptor.file.push(v);
        }

        config_set.extensions = Extensions { grpc_file_descriptor, ..Default::default() };
        Ok(config_set)
    }

    /// Performs BFS to import all nested proto files
    async fn resolve_descriptors(
        &self,
        mut descriptors: HashMap<String, FileDescriptorProto>,
        proto_path: String,
    ) -> anyhow::Result<HashMap<String, FileDescriptorProto>> {
        let parent_proto = self.read_proto(&proto_path).await?;
        let mut queue = VecDeque::new();
        queue.push_back(parent_proto.clone());

        while let Some(file) = queue.pop_front() {
            for import in file.dependency.iter() {
                let proto = self.read_proto(import).await?;
                if descriptors.get(import).is_none() {
                    queue.push_back(proto.clone());
                    descriptors.insert(import.clone(), proto);
                }
            }
        }

        descriptors.insert(proto_path, parent_proto);

        Ok(descriptors)
    }

    /// Tries to load well-known google proto files and if not found uses normal file and http IO to resolve them
    async fn read_proto(&self, path: &str) -> anyhow::Result<FileDescriptorProto> {
        let content = if let Ok(file) = GoogleFileResolver::new().open_file(path) {
            file.source()
                .context("Unable to extract content of google well-known proto file")?
                .to_string()
        } else {
            self.read_file(path).await?.content
        };

        Ok(protox_parse::parse(path, &content)?)
    }
}

#[cfg(test)]
mod test_reader {
    use std::collections::HashMap;
    use std::path::Path;

    use anyhow::{Context, Result};

    use crate::config::reader::ConfigReader;
    use crate::config::{Config, Script, ScriptOptions, Type};
    use crate::test::{init_test_runtime, path_to_file_name};

    #[tokio::test]
    async fn test_resolve() {
        // Skipping IO tests as they are covered in reader.rs
        let reader = ConfigReader::init(init_test_runtime());
        reader
            .read_proto("google/protobuf/empty.proto")
            .await
            .unwrap();
    }

    #[tokio::test]
    async fn test_nested_imports() -> Result<()> {
        let rt = init_test_runtime();
        let file_io = rt.file.clone();
        let reader = ConfigReader::init(rt);
        let helper_map = reader
            .resolve_descriptors(HashMap::new(), "src/grpc/tests/nested0.proto".into())
            .await?;
        let files = vec![
            "src/grpc/tests/nested0.proto",
            "src/grpc/tests/nested1.proto",
            "src/grpc/tests/cycle.proto",
            "src/grpc/tests/duplicate.proto",
            "src/grpc/tests/greetings.proto",
            "src/grpc/tests/news.proto",
        ];
        for path in files {
            let path_str =
                path_to_file_name(Path::new(path)).context("It must be able to extract path")?;
            let source = file_io.read(path).await?;
            let expected = protox_parse::parse(&path_str, &source)?;
            let actual = helper_map.get(&expected.name.unwrap()).unwrap();

            assert_eq!(&expected.dependency, &actual.dependency);
        }

        Ok(())
    }

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }

    #[tokio::test]
    async fn test_all() {
        let runtime = init_test_runtime();
        let file_io = runtime.file.clone();

        let mut cfg = Config::default();
        cfg.schema.query = Some("Test".to_string());
        cfg = cfg.types([("Test", Type::default())].to_vec());

        let server = start_mock_server();
        let header_serv = server.mock(|when, then| {
            when.method(httpmock::Method::GET).path("/bar.graphql");
            then.status(200).body(cfg.to_sdl());
        });

        let json = file_io.read("examples/jsonplaceholder.json").await.unwrap();

        let foo_json_server = server.mock(|when, then| {
            when.method(httpmock::Method::GET).path("/foo.json");
            then.status(200).body(json);
        });

        let port = server.port();
        let files: Vec<String> = [
            "examples/jsonplaceholder.yml", // config from local file
            format!("http://localhost:{port}/bar.graphql").as_str(), // with content-type header
            format!("http://localhost:{port}/foo.json").as_str(), // with url extension
        ]
        .iter()
        .map(|x| x.to_string())
        .collect();
        let cr = ConfigReader::init(runtime);
        let c = cr.read_all(&files).await.unwrap();
        assert_eq!(
            ["Post", "Query", "Test", "User"]
                .iter()
                .map(|i| i.to_string())
                .collect::<Vec<String>>(),
            c.types
                .keys()
                .map(|i| i.to_string())
                .collect::<Vec<String>>()
        );
        foo_json_server.assert(); // checks if the request was actually made
        header_serv.assert();
    }

    #[tokio::test]
    async fn test_local_files() {
        let runtime = init_test_runtime();

        let files: Vec<String> = [
            "examples/jsonplaceholder.yml",
            "examples/jsonplaceholder.graphql",
            "examples/jsonplaceholder.json",
        ]
        .iter()
        .map(|x| x.to_string())
        .collect();
        let cr = ConfigReader::init(runtime);
        let c = cr.read_all(&files).await.unwrap();
        pretty_assertions::assert_eq!(
            ["Post", "Query", "User"]
                .iter()
                .map(|i| i.to_string())
                .collect::<Vec<String>>(),
            c.types
                .keys()
                .map(|i| i.to_string())
                .collect::<Vec<String>>()
        );
    }

    #[tokio::test]
    async fn test_script_loader() {
        let runtime = init_test_runtime();

        let file_io = runtime.file.clone();

        let reader = ConfigReader::init(runtime);
        let config = reader
            .read("examples/jsonplaceholder_script.graphql")
            .await
            .unwrap();

        let file = ScriptOptions {
            src: file_io.read("examples/scripts/echo.js").await.unwrap(),
            timeout: None,
        };
        pretty_assertions::assert_eq!(config.server.script, Some(Script::File(file)),);
    }
}
