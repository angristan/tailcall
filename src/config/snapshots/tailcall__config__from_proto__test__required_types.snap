---
source: src/config/from_proto.rs
expression: cfg
---
schema @server @upstream {
}

type PersonRequiredFields {
  email: String
  id: Int!
  name: String!
  phone: [PhoneNumberRequiredFields]
}

type PhoneNumberRequiredFields {
  number: String!
  type: String
}
